infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/secret/SecretAsContainerResourceProvisioner.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision.secret;
+
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toMap;
+
+import com.google.common.annotations.Beta;
+import io.fabric8.kubernetes.api.model.LabelSelector;
+import io.fabric8.kubernetes.api.model.LabelSelectorBuilder;
+import io.fabric8.kubernetes.api.model.Secret;
+import java.util.Arrays;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespace;
+
+/**
+ * Finds secrets with specific labels in namespace, and mount their values as file or environment
+ * variable into workspace containers. Secrets annotated with "che.eclipse.org/mount-as=env" are
+ * mount as env variables, env name is read from "che.eclipse.org/env-name" annotation. Secrets
+ * which having "che.eclipse.org/mount-as=file" are mounted as file in the folder specified by
+ * "che.eclipse.org/mount-path" annotation. Refer to che docs for concrete examples.
+ */
+@Beta
+@Singleton
+public class SecretAsContainerResourceProvisioner<E extends KubernetesEnvironment> {
+
+  static final String ANNOTATION_PREFIX = "che.eclipse.org";
+  static final String ANNOTATION_MOUNT_AS = ANNOTATION_PREFIX + "/" + "mount-as";
+  private final FileSecretApplier fileSecretApplier;
+  private final EnvironmentVariableSecretApplier environmentVariableSecretApplier;
+
+  private final Map<String, String> secretLabels;
+
+  @Inject
+  public SecretAsContainerResourceProvisioner(
+      FileSecretApplier fileSecretApplier,
+      EnvironmentVariableSecretApplier environmentVariableSecretApplier,
+      @Named("che.workspace.provision.secret.labels") String[] labels) {
+    this.fileSecretApplier = fileSecretApplier;
+    this.environmentVariableSecretApplier = environmentVariableSecretApplier;
+    this.secretLabels =
+        Arrays.stream(labels)
+            .map(item -> item.split("=", 2))
+            .collect(toMap(p -> p[0], p -> p.length == 1 ? "" : p[1]));
```

Comment:
are you sure by this? Can `labels` be null? Is it ensured to be in correct format for this split?

Reply:
>Can labels be null? 
Nope, it's not `@Nullable` so injector will require it

> Is it ensured to be in correct format for this split?
It's not necessary. If it is not contain `=` we will just look for label with given key.